### PR TITLE
Upmcfhir 270 text support

### PR DESCRIPTION
**Pull request by mmccall**
#### What does this PR do?

Updates the EMF util to return additional sections, and adds in support for the 'text' object on Immunizations. You have to pass the bundleContext into the function to get it to work through, so we can open another ticket to go through and add that parameter pass into each function to support originalText where appropriate, this is just for Immz.

#### Related JIRA tickets:

[UPMCFHIR-270](https://jira.amida.com/browse/UPMCFHIR-270)

#### How should this be manually tested?

Look at the file and see if there is a text field there now for Immz encoding.

#### Background/Context:

Need to preserve original text in parsing some fields that aren't encoded well in Epic.

#### New JIRA tickets for clinical review required?

No

### CODE REVIEW

File: src/main/java/tr/com/srdc/cda2fhir/transform/DataTypesTransformerImpl.java [hunk 1]
```diff
@@ -242,13 +242,28 @@ public BooleanType tBL2Boolean(BL bl) {
...
+						String key = value.substring(1);
+						return idedAnnotations.get(key);
+					}
+
+				}
+			}
+
+			// If not fall back on text value.
+			String originalText = ed.getText();
+			if (!originalText.equals("")) {
```
Comment (5t33) [inline src/main/java/tr/com/srdc/cda2fhir/transform/DataTypesTransformerImpl.java:263]: Maybe we can strip whitespace here?

File: src/main/java/tr/com/srdc/cda2fhir/util/EMFUtil.java [hunk 2]
```diff
@@ -31,7 +35,8 @@ static private void putReferences(FeatureMap featureMap, Map<String, String> res
 			EStructuralFeature feature = entry.getEStructuralFeature();
 			if (feature instanceof EReference) {
 				AnyType anyType = (AnyType) entry.getValue();
-				if ("content".equalsIgnoreCase(feature.getName())) {
+
+				if (Arrays.stream(supportedTypes).anyMatch(feature.getName().toLowerCase()::equals)) {
```
Comment (5t33) [inline src/main/java/tr/com/srdc/cda2fhir/util/EMFUtil.java:39]: What type of magic is this?

**Review by 5t33 — COMMENTED** [review]
I'm having trouble seeing what the actual output will be from the html when using this. If the reference is, for example:
```
<tbody>
  <tr ID="immunization15">
     <td ID="immunization15Name">Influenza</td>
        <td>
           <content>10/31/2018</content>
        </td>
      <td />
   </tr>
 </tbody>
```
Then in the text of the codeable concept, what will be added to the immunization codeable concept? The entire HTML element? Or just the contents of the <td> tag? and what about the content?

File: src/main/java/tr/com/srdc/cda2fhir/util/EMFUtil.java (hunk 2 above)
Comment (mmccall) [reply src/main/java/tr/com/srdc/cda2fhir/util/EMFUtil.java:39]: stackoverflow "how do I work with arrays in java" magic.  basically it returns more than just "content" HTML tags now, it's those 3 in the array.

**Comment by mmccall** [comment]
> I'm having trouble seeing what the actual output will be from the html when using this. If the reference is, for example:
> 
> ```
> <tbody>
>   <tr ID="immunization15">
>      <td ID="immunization15Name">Influenza</td>
>         <td>
>            <content>10/31/2018</content>
>         </td>
>       <td />
>    </tr>
>  </tbody>
> ```
> Then in the text of the codeable concept, what will be added to the immunization codeable concept? The entire HTML element? Or just the contents of the tag? and what about the content?

The EMF util returns key/value pairs and attaches them as bundle metadata, but it only supports a few types, td being one but tr not. So the EMF util would add a key of immunization15Name, and a value of Influenza for this code. Later, in resourcetransformerImpl, you would need to pass that bundle metadata into the function that builds the codeable concepts to let it go find the original text reference values. It took me a long time to figure it out, but once you get it, it isn't that crazy.

File: src/main/java/tr/com/srdc/cda2fhir/util/EMFUtil.java (hunk 2 above)
Comment (5t33) [reply src/main/java/tr/com/srdc/cda2fhir/util/EMFUtil.java:39]: stack overflow ftw

**Comment by 5t33** [comment]
Got it. Makes sense. I'll probably approve it in a second once I walk through the code a bit more.

**Comment by 5t33** [comment]
The only issue I see with this is that the content field doesn't have an ID... but it's within a TD field that does have an ID. So We need some way to map an ID to different tag contents, e.g. Map< id, Map<field-type, value>>

**Comment by mmccall** [comment]
> The only issue I see with this is that the content field doesn't have an ID... but it's within a TD field that does have an ID. So We need some way to map an ID to different tag contents, e.g. Map< id, Map<field-type, value>>

Yeah, if we need to pull something that isn't just the text value of an entry with an id, that would be a 1-off for UPMC. We can double check that when we implement the rest of these, but I doubt it's required.

**Review by 5t33 — APPROVED** [review]
LGTM

**Pull request merged**